File: solution/backtracking/15650/main.py
# Authored by : tony9402
# Co-authored by : -
# Link : http://boj.kr/afca9b0c42354f99a6578ab92e0de0a3
import sys

def input():
    return sys.stdin.readline().rstrip()

N, M = map(int, input().split())
choose = [ 0 for _ in range(10) ]
used   = [ 0 for _ in range(10) ]

def dfs(idx, cnt):
    global N, M
    if cnt == M:
        for idx in range(cnt):
            print(choose[idx], end=' ')
        print()
        return

    for i in range(idx, N + 1):
        if used[i]:
            continue
        used[i] = 1
        choose[cnt] = i
        dfs(i + 1, cnt + 1)
        used[i] = 0

dfs(1, 0)

File: solution/backtracking/15650/main.cpp
// Authored by : tony9402
// Co-authored by : -
// Link : http://boj.kr/dc4a2e040e7d406b875abaa4d969bd50
#include<bits/stdc++.h>

using namespace std;

int choose[10], N, M;
bool used[10];

void dfs(int pre, int cnt) {
    if(cnt == M) {
        for(int i=0;i<M;i++) {
            cout << choose[i] << ' ';
        }
        cout << '\n';
        return;
    }

    for(int i=pre + 1;i<=N;i++) {
        if(used[i]) continue;
        used[i] = true;
        choose[cnt] = i;
        dfs(i, cnt + 1);
        used[i] = false;
    }
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);

    cin >> N >> M;
    dfs(0, 0);

    return 0;
}

